File: android/capacitor-intercom/src/main/java/io/stewan/capacitor/intercom/IntercomPlugin.java
package io.stewan.capacitor.intercom;

import com.getcapacitor.Config;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import io.intercom.android.sdk.Intercom;
import io.intercom.android.sdk.identity.Registration;

@NativePlugin()
public class IntercomPlugin extends Plugin {
    public static final String CONFIG_KEY_PREFIX = "plugins.IntercomPlugin.android-";

    @Override()
    public void load() {
        //
        // get config
        String apiKey = Config.getString(CONFIG_KEY_PREFIX + "apiKey", "ADD_IN_CAPACITOR_CONFIG_JSON");
        String appId = Config.getString(CONFIG_KEY_PREFIX + "appId", "ADD_IN_CAPACITOR_CONFIG_JSON");

        //
        // init intercom sdk
        Intercom.initialize(this.getActivity().getApplication(), apiKey, appId);

        //
        // load parent
        super.load();
    }


    @PluginMethod()
    public void registerIdentifiedUser(PluginCall call) {
        String email = call.getString("email");
        String userId = call.getString("userId");

        Registration registration = new Registration();

        if (email != null && email.length() > 0) {
            registration = registration.withEmail(email);
        }
        if (userId != null && userId.length() > 0) {
            registration = registration.withUserId(userId);
        }
        Intercom.client().registerIdentifiedUser(registration);
        call.success();
    }

    @PluginMethod()
    public void registerUnidentifiedUser(PluginCall call) {
        Intercom.client().registerUnidentifiedUser();
        call.success();
    }

    @PluginMethod()
    public void logout(PluginCall call) {
        Intercom.client().logout();
        call.success();
    }

    @PluginMethod()
    public void logEvent(PluginCall call) {
        String eventName = call.getString("name");
        Map<String, Object> metaData = mapFromJSON(call.getObject("data"));

        if (metaData == null) {
            Intercom.client().logEvent(eventName);
        } else {
            Intercom.client().logEvent(eventName, metaData);
        }

        call.success();
    }

    @PluginMethod()
    public void displayMessenger(PluginCall call) {
        Intercom.client().displayMessenger();
        call.success();
    }

    @PluginMethod()
    public void displayMessageComposer(PluginCall call) {
        Intercom.client().displayMessageComposer();
        call.success();
    }

    @PluginMethod()
    public void displayHelpCenter(PluginCall call) {
        Intercom.client().displayHelpCenter();
        call.success();
    }

    @PluginMethod()
    public void hideMessenger(PluginCall call) {
        Intercom.client().hideMessenger();
        call.success();
    }

    @PluginMethod()
    public void displayLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.VISIBLE);
        call.success();
    }

    @PluginMethod()
    public void hideLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.GONE);
        call.success();
    }

 
    @PluginMethod()
    public void setBottomPadding(PluginCall call) {
        Integer bottomPadding = call.getInt("bottomPaddiing");
        Intercom.client().setBottomPadding(bottomPadding);
        call.success();
    }


    private static Map<String, Object> mapFromJSON(JSONObject jsonObject) {
        if (jsonObject == null) {
            return null;
        }
        Map<String, Object> map = new HashMap<>();
        Iterator<String> keysIter = jsonObject.keys();
        while (keysIter.hasNext()) {
            String key = keysIter.next();
            Object value = getObject(jsonObject.opt(key));
            if (value != null) {
                map.put(key, value);
            }
        }
        return map;
    }

    private static Object getObject(Object value) {
        if (value instanceof JSONObject) {
            value = mapFromJSON((JSONObject) value);
        } else if (value instanceof JSONArray) {
            value = listFromJSON((JSONArray) value);
        }
        return value;
    }

    private static List<Object> listFromJSON(JSONArray jsonArray) {
        List<Object> list = new ArrayList<>();
        for (int i = 0, count = jsonArray.length(); i < count; i++) {
            Object value = getObject(jsonArray.opt(i));
            if (value != null) {
                list.add(value);
            }
        }
        return list;
    }
}

File: ios/Plugin/IntercomPlugin.swift
import Foundation
import Capacitor
import Intercom

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(IntercomPlugin)
public class IntercomPlugin: CAPPlugin {

    private var opened = false;
    private var messages = Intercom.unreadConversationCount();
    
    public override func load() {
        let apiKey = getConfig().getString("iosApiKey") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        let appId = getConfig().getString("iosAppId") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        Intercom.setApiKey(apiKey, forAppId: appId)

#if DEBUG
        Intercom.enableLogging()
#endif

        NotificationCenter.default.addObserver(self, selector: #selector(self.didRegisterWithToken(notification:)), name: Notification.Name.capacitorDidRegisterForRemoteNotifications, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.addUnreadConversationCountListener(notification:)), name: NSNotification.Name.IntercomUnreadConversationCountDidChange, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.showMessenger(_:)), name: NSNotification.Name.IntercomWindowWillShow, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.closeMessenger(_:)), name: NSNotification.Name.IntercomWindowDidHide, object: nil)
    }

    @objc func getUnreadConversationCount(_ call: CAPPluginCall) {
        call.resolve(["unreadMessages": Intercom.unreadConversationCount()])
    }

    @objc func getChatOpenedStatus(_ call: CAPPluginCall) {
        call.resolve(["isChatOpen": opened]);
    }

    @objc func showMessenger(_ call: CAPPluginCall) {
        // print("showMessenger called")
        opened = true
        messages = 0
    }

    @objc func closeMessenger(_ call: CAPPluginCall) {
        // print("hide notification called")
        opened = false;
    }

    @objc func addUnreadConversationCountListener(notification: NSNotification) {
        print("initial count ", messages)
        let unreadCount = Intercom.unreadConversationCount()
        print("after count ", unreadCount)
        messages = messages + unreadCount
        notifyListeners("IntercomUnreadConversationCountDidChange", data: ["unreadMessages": messages])
        // call.resolve()
    }

    @objc func didRegisterWithToken(notification: NSNotification) {
        guard let deviceToken = notification.object as? Data else {
            return
        }
        Intercom.setDeviceToken(deviceToken)
    }

    @objc func registerIdentifiedUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let email = call.getString("email")
        let attributes = ICMUserAttributes()

        if ((email) != nil) {
            attributes.email = email
            Intercom.loginUser(with: attributes) { result in
                switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                }
            }
        }

        if ((userId) != nil) {
            attributes.userId = userId
            Intercom.loginUser(with: attributes) { result in
                switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                }
            }
        }
    }

    @objc func registerUnidentifiedUser(_ call: CAPPluginCall) {
        Intercom.loginUnidentifiedUser()
        call.resolve()
    }

    @objc func updateUser(_ call: CAPPluginCall) {
        let userAttributes = ICMUserAttributes()
        let userId = call.getString("userId")
        if (userId != nil) {
            userAttributes.userId = userId
        }
        let email = call.getString("email")
        if (email != nil) {
            userAttributes.email = email
        }
        let name = call.getString("name")
        if (name != nil) {
            userAttributes.name = name
        }
        let phone = call.getString("phone")
        if (phone != nil) {
            userAttributes.phone = phone
        }
        let languageOverride = call.getString("languageOverride")
        if (languageOverride != nil) {
            userAttributes.languageOverride = languageOverride
        }
        let customAttributes = call.getObject("customAttributes")
        userAttributes.customAttributes = customAttributes
        Intercom.updateUser(with: userAttributes)
        call.resolve()
    }

    @objc func logout(_ call: CAPPluginCall) {
        Intercom.logout()
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name("IntercomUnreadConversationCountDidChange"), object: nil);
        call.resolve()
    }

    @objc func logEvent(_ call: CAPPluginCall) {
        let eventName = call.getString("name")
        let metaData = call.getObject("data")

        if (eventName != nil && metaData != nil) {
            Intercom.logEvent(withName: eventName!, metaData: metaData!)

        }else if (eventName != nil) {
            Intercom.logEvent(withName: eventName!)
        }

        call.resolve()
    }

    @objc func displayMessenger(_ call: CAPPluginCall) {
        Intercom.present();
        call.resolve()
    }

    @objc func displayMessageComposer(_ call: CAPPluginCall) {
        guard let initialMessage = call.getString("message") else {
            call.reject("Enter an initial message")
            return
        }
        Intercom.presentMessageComposer(initialMessage);
        call.resolve()
    }

    @objc func displayHelpCenter(_ call: CAPPluginCall) {
        Intercom.presentHelpCenter()
        call.resolve()
    }

    @objc func hideMessenger(_ call: CAPPluginCall) {
        print("hide messenger called")
        Intercom.hide()
        call.resolve()
    }

    @objc func displayLauncher(_ call: CAPPluginCall) {
        Intercom.setLauncherVisible(true)
        call.resolve()
    }

    @objc func hideLauncher(_ call: CAPPluginCall) {
        Intercom.setLauncherVisible(false)
        call.resolve()
    }

    @objc func displayInAppMessages(_ call: CAPPluginCall) {
        Intercom.setInAppMessagesVisible(true)
        call.resolve()
    }

    @objc func hideInAppMessages(_ call: CAPPluginCall) {
        print("hide in app messages")
        Intercom.setInAppMessagesVisible(false)
        call.resolve()
    }

    @objc func displayCarousel(_ call: CAPPluginCall) {
        if let carouselId = call.getString("carouselId") {
            Intercom.presentCarousel(carouselId)
            call.resolve()
        }else{
            call.reject("carouselId not provided to displayCarousel.")
        }
    }

    @objc func setUserHash(_ call: CAPPluginCall) {
        let hmac = call.getString("hmac")

        if (hmac != nil) {
            Intercom.setUserHash(hmac!)
            call.resolve()
            print("hmac sent to intercom")
        }else{
            call.reject("No hmac found. Read intercom docs and generate it.")
        }
    }

    @objc func setBottomPadding(_ call: CAPPluginCall) {

        if let value = call.getString("value"),
           let number = NumberFormatter().number(from: value) {

            Intercom.setBottomPadding(CGFloat(truncating: number))
            call.resolve()
            print("set bottom padding")
        } else {
            call.reject("Enter a value for padding bottom")
        }
    }

    @objc func displayArticle(_ call: CAPPluginCall) {
        if let articleId = call.getString("articleId") {
            Intercom.presentArticle(articleId)
            call.resolve()
        } else {
            call.reject("articleId not provided to presentArticle.")
        }
    }
}
